compiler/rustc_typeck/src/check/demand.rs

```diff
@@ -577,12 +578,19 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                 // We have `&T`, check if what was expected was `T`. If so,
                 // we may want to suggest removing a `&`.
                 if sm.is_imported(expr.span) {
-                    if let Ok(src) = sm.span_to_snippet(sp) {
-                        if let Some(src) = src.strip_prefix('&') {
+                    // Go through the spans from which this span was expanded,
+                    // and find the one that's pointing inside `sp`.
+                    //
+                    // E.g. for `&format!("")`, where we want the span to the
+                    // `format!()` invocation instead of its expansion.
+                    if let Some(call_span) =
+                        iter::successors(Some(expr.span), |s| s.parent()).find(|&s| sp.contains(s))
+                    {
+                        if let Ok(code) = sm.span_to_snippet(call_span) {
                             return Some((
                                 sp,
                                 "consider removing the borrow",
-                                src.to_string(),
+                                code,
```

Comment:
In another diagnostic, I emitted (potentially multiple) suggestions for removing the parts that need to be removed, instead of a single suggestion that contains a copy of (part of) the original source. Is there a guideline for which to prefer?

Basically:
A. "replace `&mut (format!(""))` by `format!("")`
B. "remove `&mut (` and remove `)`"

Most diagnostics seem to use A, which surprises me a bit. The suggestion for A can get quite large for large expressions.

Reply:
The later didn't exist until "recently" (2019?), so there's a historical component. There's also a bug that needs to be fixed on rustc's side to allow rustfix apply multipart suggestions, which it can't today.

Reply:
Thanks!

> There's also a bug that needs to be fixed on rustc's side to allow rustfix apply multipart suggestions, which it can't today.

That should be fixed by https://github.com/rust-lang/rustfix/pull/195

Reply:
Awesome! That should also allow us to add a `// run-rustfix` to a bunch of tests that don't have it today because of that 😄